### PR TITLE
[Refactor] refactor get_subset_by_classes in dataloader for training with empty-GT images

### LINKED ISSUE #2965: Custom dataset question - classes specification
**Checklist**
1. I have searched related issues but cannot get the expected help.
2. The bug has not been fixed in the latest version.

**Describe the bug**
Cannot modify classes in config.py. Have to modify hard code. And if you modify the config, the training will start with a evaluation. Only happens in mmdet v2.0.
**Reproduction**
1. What command or script did you run?
```
./tools/dist_train.sh configs/cascade_rcnn/cascade_rcnn_x101_64x4d_fpn.py  2  --work-dir /data/cascade_rcnn_x101_64x4d_fpn
```
2. Did you make any modifications on the code or config? Did you understand what you have modified?
```
_base_ = './cascade_rcnn_r50_fpn_1x_coco.py'
model = dict(
    type='CascadeRCNN',
    pretrained='open-mmlab://resnext101_64x4d',
    backbone=dict(
        type='ResNeXt',
        depth=101,
        groups=64,
        base_width=4,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=True),
        style='pytorch'),
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            scales=[4],
            ratios=[0.5, 1.0, 2.0],
            strides=[4, 8, 16, 32, 64]),
        bbox_coder=dict(
            type='DeltaXYWHBBoxCoder',
            target_means=[.0, .0, .0, .0],
            target_stds=[1.0, 1.0, 1.0, 1.0]),
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
        loss_bbox=dict(type='SmoothL1Loss', beta=1.0 / 9.0, loss_weight=1.0)),

    roi_head=dict(
        type='CascadeRoIHead',
        num_stages=3,
        stage_loss_weights=[1, 0.5, 0.25],
        bbox_roi_extractor=dict(
            type='SingleRoIExtractor',
            roi_layer=dict(type='RoIAlign', out_size=7, sample_num=0),
            out_channels=256,
            featmap_strides=[4, 8, 16, 32]),
        bbox_head=[
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=10,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0., 0., 0., 0.],
                    target_stds=[0.1, 0.1, 0.2, 0.2]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0,
                               loss_weight=1.0)),
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=10,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0., 0., 0., 0.],
                    target_stds=[0.05, 0.05, 0.1, 0.1]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0,
                               loss_weight=1.0)),
            dict(
                type='Shared2FCBBoxHead',
                in_channels=256,
                fc_out_channels=1024,
                roi_feat_size=7,
                num_classes=10,
                bbox_coder=dict(
                    type='DeltaXYWHBBoxCoder',
                    target_means=[0., 0., 0., 0.],
                    target_stds=[0.033, 0.033, 0.067, 0.067]),
                reg_class_agnostic=True,
                loss_cls=dict(
                    type='CrossEntropyLoss',
                    use_sigmoid=False,
                    loss_weight=1.0),
                loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0))
        ]))
# dataset settings
dataset_type = 'CocoDataset'
classes = ('pedestrian', 'people', 'bicycle', 'car', 'van', 'truck', 'tricycle', 'awning-tricycle', 'bus', 'motor')
data_root = '/data/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='MinIoURandomCrop',
         min_ious=(0.1, 0.3, 0.5, 0.7, 0.9),
         min_crop_size=0.3),
    dict(type='Resize', img_scale=(1520, 1002), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1520, 1002),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    samples_per_gpu=1,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        classes=classes,
        ann_file=data_root + 'xxx/train.json',
        img_prefix=data_root + 'xxx/images/',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        classes=classes,
        ann_file=data_root + 'xxx/val.json',
        img_prefix=data_root + 'xxx/images/',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        classes=classes,
        ann_file=data_root + 'xxx/val.json',
        img_prefix=data_root + 'xxx/images/',
        pipeline=test_pipeline))
evaluation = dict(interval=1, metric='bbox')
optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0001)
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[12, 18, 26])
total_epochs = 30

```
3. What dataset did you use?

**Environment**
```
sys.platform: linux
Python: 3.7.7 (default, May  7 2020, 21:25:33) [GCC 7.3.0]
CUDA available: True
CUDA_HOME: /usr/local/cuda-10.1
NVCC: Cuda compilation tools, release 10.1, V10.1.243
GPU 0,1: GeForce GTX 1080 Ti
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.3.1
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2019.0.4 Product Build 20190411 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.3
  - Magma 2.5.1
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF, 

TorchVision: 0.4.2
OpenCV: 4.2.0
MMCV: 0.5.9
MMDetection: 2.0.0+204f751
MMDetection Compiler: GCC 7.5
MMDetection CUDA Compiler: 10.1
```

**Error traceback**
If applicable, paste the error trackback here.
```
Traceback (most recent call last):
  File "./tools/train.py", line 161, in <module>
    main()
  File "./tools/train.py", line 157, in main
    meta=meta)
  File "/home/omnisky/yuweiping/mmdetection/mmdet/apis/train.py", line 179, in train_detector
    runner.run(data_loaders, cfg.workflow, cfg.total_epochs)
  File "/home/omnisky/anaconda3/envs/mmdet2/lib/python3.6/site-packages/mmcv/runner/runner.py", line 384, in run
    epoch_runner(data_loaders[i], **kwargs)
  File "/home/omnisky/anaconda3/envs/mmdet2/lib/python3.6/site-packages/mmcv/runner/runner.py", line 293, in train
    self.call_hook('after_train_epoch')
  File "/home/omnisky/anaconda3/envs/mmdet2/lib/python3.6/site-packages/mmcv/runner/runner.py", line 245, in call_hook
    getattr(hook, fn_name)(self)
  File "/home/omnisky/yuweiping/mmdetection/mmdet/core/evaluation/eval_hooks.py", line 74, in after_train_epoch
    self.evaluate(runner, results)
  File "/home/omnisky/yuweiping/mmdetection/mmdet/core/evaluation/eval_hooks.py", line 32, in evaluate
    results, logger=runner.logger, **self.eval_kwargs)
  File "/home/omnisky/yuweiping/mmdetection/mmdet/datasets/coco.py", line 343, in evaluate
    result_files, tmp_dir = self.format_results(results, jsonfile_prefix)
  File "/home/omnisky/yuweiping/mmdetection/mmdet/datasets/coco.py", line 306, in format_results
    result_files = self.results2json(results, jsonfile_prefix)
  File "/home/omnisky/yuweiping/mmdetection/mmdet/datasets/coco.py", line 240, in results2json
    if isinstance(results[0], list):
IndexError: list index out of range

```

### LINKED ISSUE #3545: Filtering out the empty images by default
I notice that on the customized dataset, the data loader by default will filter out empty images (image without annotations) on the testing set during testing. Is there a particular reason for doing so? Thanks!

```
    CLASSES = None

    def __init__(self,
                 ann_file,
                 pipeline,
                 classes=None,
                 data_root=None,
                 img_prefix='',
                 seg_prefix=None,
                 proposal_file=None,
                 test_mode=False,
                 **filter_empty_gt=True):**
```

### PR DESCRIPTION
**Pull request by ZwwWayne**
This PR fixes the issue of training with empty-GT images when specifying classes, continuing on https://github.com/open-mmlab/mmdetection/pull/3133.

In the original implementation, `get_subset_by_classes` is not related to the flag `self.filter_empty_gt`. This brings ambiguous behavior and potential bugs when specifying dataset classes. More specifically, `CocoDataset` filters empty gt images in two places, one is in `_filter_img`, the other one is in `get_subset_by_classes`. Other datasets only filter empty gt images in `get_subset_by_classes`, but this will be triggered by specifying classes rather than `self.filter_empty_gt=True`. 

The goal of this PR is to make the behavior of `filter empty GT` independent to `specifying classes`. After this PR, if `filter_empty_gt=False`, no matter whether the classes are specified in a dataset, the dataset will use all the images in the annotations. If `filter_empty_gt=True`, no matter whether the classes are specified, the dataset will check the images and filter out images containing no GT boxes.

This PR also changes the logic during test_mode. Previously, the dataset will filter out empty-GT images no matter whether it is in test mode. After this PR, it only filters out images in training, i.e., if `test_mode=True`, it will test all the images given in the annotations. This behavior is more commonly need as mentioned in #3545, and is more consistent with before that the dataset will not filter out small images in test mode. Therefore, the users should be responsible for the data filtering/cleaning process for the test dataset.

Fix https://github.com/open-mmlab/mmdetection/issues/3023, Fix #2965, Fix #3545

### CODE REVIEW

File: mmdet/datasets/coco.py [hunk 1]
```diff
@@ -88,38 +88,22 @@ def _filter_imgs(self, min_size=32):
         """Filter images too small or without ground truths."""
         valid_inds = []
         ids_with_ann = set(_['image_id'] for _ in self.coco.anns.values())
+        ids_in_cat = set()
```
Comment (Johnson-Wang) [inline mmdet/datasets/coco.py:91]: Some comments may be necessary.

File: mmdet/datasets/custom.py [hunk 2]
```diff
@@ -237,12 +238,13 @@ def get_classes(cls, classes=None):
                 string, take it as a file name. The file contains the name of
                 classes where each line contains one class name. If classes is
                 a tuple or list, override the CLASSES defined by the dataset.
+
+        Returns:
+            tuple[str]: Names of categories of the dataset.
```
Comment (Johnson-Wang) [inline mmdet/datasets/custom.py:243]: We may add a conversion from list to tuple before `return` since there are some in list.

**Review by Johnson-Wang — APPROVED** [review]

File: mmdet/datasets/coco.py [hunk 3]
```diff
@@ -87,39 +87,27 @@ def get_cat_ids(self, idx):
     def _filter_imgs(self, min_size=32):
         """Filter images too small or without ground truths."""
         valid_inds = []
+        # obtain images that contain annotation
         ids_with_ann = set(_['image_id'] for _ in self.coco.anns.values())
+        # obtain images that contain annotations of the required categories
```
Comment (OceanPang) [inline mmdet/datasets/coco.py:92]: It's not necessary to do the filtering here. 
We can directly do that in `load_annotations`,
`self.cat_ids = self.coco.get_cat_ids(cat_names=self.CLASSES)`
`self.img_ids = self.coco.get_img_ids(cat_ids=self.cat_ids)`

File: mmdet/datasets/custom.py [hunk 4]
```diff
@@ -80,23 +82,21 @@ def __init__(self,
...
-        # filter data infos if classes are customized
-        if self.custom_classes:
-            self.data_infos = self.get_subset_by_classes()
 
         if self.proposal_file is not None:
             self.proposals = self.load_proposals(self.proposal_file)
         else:
             self.proposals = None
-        # filter images too small
+
```
Comment (OceanPang) [inline mmdet/datasets/custom.py:90]: Considering implement `get_subset_by_cls` in other Dataset.

File: mmdet/datasets/coco.py (hunk 3 above)
Comment (ZwwWayne) [reply mmdet/datasets/coco.py:92]: self.coco.get_img_ids will only return images that contain all the categories specified by the `cat_ids`.

**Comment by OceanPang** [comment]
@ZwwWayne  We should also implement `get_subset_by_classes` during testing time, which is different vs. the mentioned issue. 
For example, if you want to train only the `person` class in COCO and validate so, current APIs do not satisfy the needs.

**Review by OceanPang — APPROVED** [review]

**Comment by hellock** [comment]
Task linked: [CU-4hu3hu PR #3695](https://app.clickup.com/t/4hu3hu)

**Pull request merged**